### ISSUE TITLE
Reorganize attribute documentation

### ISSUE THREAD

**Issue by Ininterrompue** [JuliaPlots/PlotDocs.jl#323]
The current documentation for attributes could use some refreshing.

1. Most attributes cannot be searched directly in the search bar (known problem). Ideally the way that attributes are documented should be done in a way where they *can* be searched. Maybe we should replace the table with something closer to the API page?
2. Attributes are not organized well. Organizing them into series, plot, subplot, and axis is something I have found very confusing. 
   * In particular, typing something like `plotattr(:Subplot)` in the REPL outputs a large list of attributes. This hinders discoverability and accessibility. In my experience, I have always had to look through all four tabs in the docs to look for an attribute.
   * What would be better for `plotattr(:Subplot)` is to reorganize the categories into things like Annotation, Background, Foreground, Colorbar, Legend, Title, as well as a few others. A lot of the attribute names in this list heavily suggest this anyway. This way, when one needs to find attributes related to the legend, there is a clearly labeled tab in the docs for that, or typing e.g. `plotattr(:Legend)` would yield all of its relevant attributes.
3. The types required for the attributes are not always intuitive. For example, `colorbar_ticks` could use some explaining on the different possible argument types, as `Union{Symbol, Tuple{AbstractVector{Real}, AbstractVector{AbstractString}}, AbstractVector{Real}}` is not obvious how to interpret. Also, a lot of attributes have no description. 

A documentation style like that on the API page would be able to solve these problems simply by allowing more room for the description, explanation of types, and examples. Categorizing them into titles like Colorbar, Legend, etc. would keep these pages manageable in length.

**Comment by mkborregaard** [comment]
These are good suggestions IMHO. See also https://github.com/JuliaPlots/Plots.jl/issues/676 bit unsure whether this https://github.com/JuliaPlots/Plots.jl/issues/2337 should still be open

**Comment by t-bltg** [comment]
Point 1. has finally been fixed yesterday. When searching for e.g. `show_empty_bins` a search results with https://docs.juliaplots.org/stable/generated/attributes_series/ is proposed.
You can search for attributes but also aliases.

Not sure about 2. (I don't think the proposed approach would enhance clarity).

Point 3. is a `julia` language pre-requisite and has nothing to do with `Plots`, although since beginners are likely to use `Plots` rapidly, we could maybe add a link to https://docs.julialang.org/en/v1/manual/types/#Type-Unions and a note in the attributes page.

Maybe somewhere here: https://github.com/JuliaPlots/PlotDocs.jl/blob/master/src/PlotDocs.jl#L267-L270.

**Comment by Ininterrompue** [comment]
> Point 1. has finally been fixed yesterday (#318 - #320). When searching for e.g. `show_empty_bins`, a search results pointing to https://docs.juliaplots.org/stable/generated/attributes_series/ is proposed. You can search for attributes but also aliases now.

If I search `show_empty_bins`, the relevant hits with the list of attributes are the last two. I'm not sure about the details of the search mechanism with `Documenter.jl`. This may be an issue with that package.

If I search `markersize`, the first result should point straight to the page that has the information about this attribute. But this page is far down the list, and it's not obvious which one. (It's the one called "Series Attributes") Also, if I search for its alias `ms`, it's basically at the bottom of the results.

However, if the attribute pages were prioritized more in the search results, this would help greatly with the discoverability issue. In my experience, 95% of the time I searched for something, it was an attribute.

> Not sure about 2. (I am not sure that the proposed approach would enhance clarity).

I think it would enhance clarity greatly. I'm not the only one that has been frustrated with finding attribute documentation. See [here](https://discourse.julialang.org/t/discussion-on-why-i-no-longer-recommend-julia-by-yuri-vishnevsky/81151/278), [here](https://discourse.julialang.org/t/discussion-on-why-i-no-longer-recommend-julia-by-yuri-vishnevsky/81151/279), [here](https://discourse.julialang.org/t/discussion-on-why-i-no-longer-recommend-julia-by-yuri-vishnevsky/81151/280). Also, anecdotes from my friends. I'm motivated to fix this partly because this was by far the most frustrating thing about learning Julia for me. 

I remember trying to figure out how to manipulate the legend and could not figure out which one of Series, Plot, Subplot, and Axis had the entries for legend attributes, even after many iterations. But if they were all listed under a section called "Legend Attributes," it would be obvious. Legend attributes was particularly confusing since one can have a plot of lines and points all on the same canvas, with one legend, and yet the attributes are listed under Subplot, not Plot.

> Point 3. is a `julia` language pre-requisite and has nothing to do with `Plots`, although since beginners are likely to use `Plots` rapidly, we could maybe add a link to https://docs.julialang.org/en/v1/manual/types/#Type-Unions and a note in the attributes page. Maybe somewhere around here: https://github.com/JuliaPlots/PlotDocs.jl/blob/master/src/PlotDocs.jl#L267-L270.

The type system is indeed something baked into the language itself. However, it is a good idea to explain the types and give examples of the possible values that it can take. This is especially important since beginners and many end-users do not regularly use or encounter generic code, yet these are the very people that probably want to know how to plot something. 

To give another example, `bins` has the type

```
Union{Integer, Symbol, Tuple{Integer, Integer}, AbstractVector}
```

but a lot of people don't know that `StepRange <: AbstractVector`, or may be confused by `Int` vs. `Integer`. Also, I tried to write the following code

```julia
using Plots
x = randn(10^4)
histogram(x, bins=(-1, 1))
```

but I get a `MethodError`:

```
ERROR: MethodError: no method matching _hist_edges(::Tuple{Vector{Float64}}, ::Tuple{Int64, Int64})
Closest candidates are:
  _hist_edges(::Tuple{Vararg{AbstractVector, N}}, ::Tuple{Vararg{Any, N}}) where N at C:\Users\user\.julia\packages\Plots\Hxe7H\src\recipes.jl:757
  _hist_edges(::Tuple{Vararg{AbstractVector, N}}, ::Union{Integer, Symbol, AbstractVector}) where N at C:\Users\user\.julia\packages\Plots\Hxe7H\src\recipes.jl:763
```
so it's not obvious to me how to make use of `Tuple{Integer, Integer}`.

Here is how I would write the following doc entry, leaving out `Tuple{Integer, Integer}`:

```julia
bins :: Union{Integer, Symbol, Tuple{Integer, Integer}, AbstractVector}
```
Defines the number of bins.

Can take in one of the following types:
* `Integer` – abstract supertype for all integers. Defines the approximate number of bins to aim for.
    * `bins=10` will give a histogram with approximately 10 bins, although it is not guaranteed. 
* `Symbol` – a name or label to identify an entity. Defines the auto-binning algorithm to use:
    * `:sturges` – [Sturges' rule](https://en.wikipedia.org/wiki/Histogram#Sturges'_formula)
    * `:sqrt` – [Square root rule](https://en.wikipedia.org/wiki/Histogram#Square-root_choice)
    * `:rice` – [Rice rule](https://en.wikipedia.org/wiki/Histogram#Rice_rule) 
    * `:scott` – [Scott's normal reference rule](https://en.wikipedia.org/wiki/Histogram#Scott's_normal_reference_rule)
    * `:fd` (default) – [Freedman-Diaconis' rule](https://en.wikipedia.org/wiki/Histogram#Freedman%E2%80%93Diaconis'_choice)
* `AbstractVector` – supertype for one-dimensional arrays (or array-like types). Defines a vector of values for bin edges.
    * `bins=range(-10, 10, length=21)` will give a histogram with bins starting from -10, ending at 10, and containing 21 break values, giving 20 bins.

Relevant attribute for the following series types:
* [`histogram`](https://docs.juliaplots.org/stable/api/#Plots.histogram-Tuple)
* [`histogram2d`](https://docs.juliaplots.org/stable/api/#Plots.histogram2d-Tuple)

**Comment by t-bltg** [comment]
> If I search show_empty_bins, the relevant hits with the list of attributes are the last two. I'm not sure about the details of the search mechanism with Documenter.jl. This may be an issue with that package.

Yeah well I didn't say it was perfect. The current solution is monkey patching the `domify` method from `Documenter` in order to tweak the search index when using raw html tables:

https://github.com/JuliaPlots/PlotDocs.jl/blob/48095129f658a04706481b788cc134bed65fc357/docs/make.jl#L14-L48

**Comment by t-bltg** [comment]
> If I search show_empty_bins, the relevant hits with the list of attributes are the last two. I'm not sure about the details of the search mechanism with Documenter.jl. This may be an issue with that package.

Yeah well I didn't say it was perfect. The current solution is monkey patching the `domify` method from `Documenter` in order to tweak the search index when using raw html tables:

https://github.com/JuliaPlots/PlotDocs.jl/blob/48095129f658a04706481b788cc134bed65fc357/docs/make.jl#L14-L48

I agree that it would be better to weight the search result in order to make the result appear in the first results. As you can see in [search.js](https://github.com/JuliaDocs/Documenter.jl/blob/master/assets/html/search.js#L189-L209), there is a boost field, that we could use to tweak the search order. I gave it a try but couldn't find a solution while implementing https://github.com/JuliaPlots/PlotDocs.jl/pull/318 (which in the end is just a hack).

> But if they were all listed under a section called "Legend Attributes," it would be obvious.

That could work, but that would need some effort to integrate smoothly in `Plots` and `PlotDocs`. It's "only" a matter of putting in development time, and producing something not being too convoluted and difficult to maintain because of its complexity.

> Instead, here is how I would write the following doc entry:

That looks nice indeed. Currently you can see that the description types are defined [here](https://github.com/JuliaPlots/Plots.jl/blob/f39c4e52845d221089b33945c89bbe5ec17c6c79/src/arg_desc.jl#L1-L9). If you can make it work without increasing the complexity too much I'd be happy to integrate this, but be aware that implementing a clean and generic solution for the arguments and their description might be non-trivial.

**Comment by Ininterrompue** [comment]
With the entry I wrote, I think it's a bit too complex to be included as part of a dictionary. The majority of attributes won't be as complex as `bins`, but there are still quite a few that need more than a few lines to completely document.

My idea is the tables should be replaced with pages that resemble the [API page](https://docs.juliaplots.org/stable/api/#RecipesBase.plot). The disadvantages with the tables right now is that the attributes cannot directly be searched, and the amount of information for a given attribute is softly constrained by a box. (Not an actual constraint...but the dictionary+table format strongly suggests terse descriptions) 

With the API pages, all the information I wrote can be put into an entry for `bins` and formatted just like above, and it will also be easier to search for the particular attribute in the search bar (but enhancing the search itself should be a separate issue).

One issue is how to include this information in the REPL. Right now, typing `plotattr("bins")` gives the information, but it is different than typing `>?foo` since `foo` has to be a method. Would this work if entries were placed on an API page? I understand that this is not how keywords are normally documented...

I'm willing to put in the work to recategorize all of the attributes and create separate API pages for each of them. Just want to figure out the best way to approach this.

**Comment by Ininterrompue** [comment]
One way to recategorize them is the following:

* [**Series attributes**](https://docs.juliaplots.org/stable/generated/attributes_series/) stay the same. On the API page, every entry contains a "Relevant attribute for the following series types:" list that denotes all of the series types which the attribute can modify. This will also help with the outstanding issue(?) that irrelevant attributes are ignored - the docs would explicitly show support.
* [Plot attributes](https://docs.juliaplots.org/stable/generated/attributes_plot/) is renamed to **Global attributes.** This is probably a less ambiguous word to refer to those attributes which impact the entire Plot.
* [Subplot attributes](https://docs.juliaplots.org/stable/generated/attributes_subplot/) are split up into the following categories:
    * **Annotation attributes:** `annotationcolor` `annotationfontfamily` `annotationfontsize` `annotationhalign` `annotationrotation` `annotations` `annotationvalign`
    * **Colorbar attributes:** `clims` `colorbar` `colorbar_continuous_values` `colorbar_discrete_values` `colorbar_fontfamily` `colorbar_formatter` `colorbar_scale` `colorbar_tickfontcolor` `colorbar_tickfontfamily` `colorbar_tickfonthalign` `colorbar_tickfontrotation` `colorbar_tickfontsize` `colorbar_tickfontvalign` `colorbar_ticks` `colorbar_title` `colorbar_title_location` `colorbar_titlefontcolor` `colorbar_titlefontfamily` `colorbar_titlefonthalign` `colorbar_titlefontrotation` `colorbar_titlefontsize` `colorbar_titlefontvalign`
    * **Legend attributes:** `legend_background_color` `legend_column` `legend_font` `legend_font_color` `legend_font_family` `legend_font_halign` `legend_font_pointsize` `legend_font_rotation` `legend_font_valign` `legend_foreground_color` `legend_position` `legend_title` `legend_title_font` `legend_title_font_family` `legend_title_font_halign` 
`legend_title_font_pointsize` `legend_title_font_rotation` `legend_title_font_valign`
    * **Margin attributes:** `bottom_margin` `left_margin` `margin` `right_margin` `top_margin`
    * **Title attributes:** `title` `titlefontcolor` `titlefontfamily` `titlefonthalign` `titlefontrotation` `titlefontsize` `titlefontvalign` `titlelocation`
    * **3D-specific attributes:** `camera` `projection` `projection_type`
    * **Other subplot attributes:** `aspect_ratio` `background_color_inside` `background_color_subplot` `color_palette` `extra_kwargs` `fontfamily_subplot` `foreground_color_subplot` `foreground_color_title` `framestyle` `subplot_index`
* [**Axis attributes**](https://docs.juliaplots.org/stable/generated/attributes_axis/) stay the same.

We might also want to consider making the attribute names consistent with regards to spacing, snake-case, etc. Right now, all the annotation attributes are concatenated words, the colorbar attributes are all over the place, the legend attributes are all snakecase, and the title attributes are concatenated. Also, `clims` (colorbar limits) has an official name that is different from the other colorbar attributes.

**Comment by BeastyBlacksmith** [comment]
What we shouldn't change is that there is a single source of truth.
On the other hand I don't see a reason why you couldn't generate the markdown in https://github.com/JuliaPlots/PlotDocs.jl/issues/323#issuecomment-1345472075 from the current stucture when you change the descripion string in Plots.jl.
I would probably like to a frontpage that only lists the attributes as links, such that you get the detailed information if you click on it.

**Comment by Ininterrompue** [comment]
> On the other hand I don't see a reason why you couldn't generate the markdown in [#323 (comment)](https://github.com/JuliaPlots/PlotDocs.jl/issues/323#issuecomment-1345472075) from the current stucture when you change the descripion string in Plots.jl. 

I'm not sure I follow. It is technically possible to fit all of the information in the table, but it wouldn't look good as the last column would take up too much space. There's a bunch of nested bullet points, and for some attributes, the list of relevant series types which the attribute applies to can be huge. The information really needs the entire space on the page to be readable, not just a single column.

**Comment by BeastyBlacksmith** [comment]
I'm not saying we should keep that table. We can generate from the dictionary in `Plots.jl/src/attr_descr.jl` any markdown.

**Comment by Ininterrompue** [comment]
Does the information have to be written as part of a dictionary, or can I write the markdown directly? Using `##` and `###` headers would be good enough for organization.

**Comment by BeastyBlacksmith** [comment]
In principle its fine to use markdown stings there, however we should take care not to break usage of that information in `plotattr` and docstrings.
Thats either keeping the dictionary structure or also changing how these function use that information.

**Comment by Ininterrompue** [comment]
One way is to keep the dictionary as-is, but create new pages with pure Markdown, replacing the `generated/attribute_series.md`, etc. That way `plotattr` functionality won't be broken. There'll be a disconnect between documentation on the site and documentation as seen in the REPL via `plotattr`, but I think this'll be a good start. What do you think?

**Comment by BeastyBlacksmith** [comment]
Sounds reasonable to me

**Comment by Ininterrompue** [comment]
[Example](https://github.com/Ininterrompue/PlotDocs.jl/blob/master/docs/src/Attributes/attributes_series.md) for series attributes. While creating this page, it was found that `match_dimensions` was not being generated in the table.
